### ISSUE TITLE
Technical Question: should we use a `JWT` for `AUTH_API_KEY`?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/auth#268]
At present an `AUTH_API_KEY` has the format:
```sh
88SwQDWYGmiDACVRaBGqQghzLd5jfX7YynefCJ8M/88SwQH8CEDWzWSq7PyWe7ADpjYCpmczr5zyTK4d/authdemo.fly.dev
```
> **Note**: this is _not_ a valid key. but the format is correct.

The _reason_ I wet with this and _still_ like it is because it's **human-readable**
and **_immediately_ obvious which environment** (`auth instance`) it's for; in this case: **`authdemo.fly.dev`**
So if I see this key in my `.env` file or environment variable, I know _exactly_ what it's for. 👌 
I think this has a pretty major advantage in terms of Developer Experience ... 💭 

But equally it has a disadvantage: no embedded info like `expiry`.
We outlined the benefits of `JWTs` in our mega-popular doc: https://github.com/dwyl/learn-json-web-tokens
And we are _using_ them in `auth` for session tokens:

<img width="892" alt="dwyl-mvp-jwt" src="https://user-images.githubusercontent.com/194400/217531448-c911ebc4-99fc-4daa-a8d2-a0c883a6e7f9.png">

https://github.com/dwyl/auth/blob/83c286a24d47bd087ba41925a740d4030cf95e19/lib/auth_web/controllers/auth_controller.ex#L200

So my question is: should the _next_ version of `auth` (and `auth_plug`) use a `JWT` as the `AUTH_API_KEY`?
Will having a `JWT` for the `AUTH_API_KEY` and a ***different*** `JWT` for session token be confusing to devs? 🤷‍♂️ 
Are we _missing_ something? 

Please let me know your thoughts ... 💭 🙏

**Comment by LuchoTurtle** [comment]
I don't think it makes much sense to have info like `expiry` in the `AUTH_API_KEY`. In my opinion, the key should be a unique string that the application uses to guarantee that the application using it is genuine in the "`auth` ecosystem".

However, the `JWT` session token that is returned to the browser and to the user should certainly have such kind of metadata. After all, `JWT` is just a way of encoding claims between two parties into a string. This session token should certainly be `JWT` but I fail to see the usefulness of having metadata info in a **key** that is used as env variable - it should be present in the **session token** `JWT`.

A similar thing happens when working with [OpenID Connect](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc). 

If I have a *client*, a *server* and an *identity provider* like [Keycloak](https://www.keycloak.org/), for example, the server uses in deployment a **key as env variable** that the **identity provider issues**, so it *knows* the server is legitimate. This key doesn't need to be a JWT. 

However, if I were to use an OIDC (OpenID Connect) flow for users to authenticate in the server...

<img width="831" alt="image" src="https://user-images.githubusercontent.com/17494745/217810717-247f8106-e4ff-4537-8e14-415718024b4b.png">

the interactions between the *client* and the *server* are made through JWT tokens. But the *web server securely communicates with the identity provider to check if the issued [ID tokens and Access Tokens](https://oauth.net/id-tokens-vs-access-tokens/#:~:text=Access%20tokens%20are%20what%20the,read%20by%20the%20OAuth%20client.) (part of the OIDC authentication flow) are legitimate. **And the only reason the identity provider *is the provider for the server*, is because it gave the server "a key" so it knows it's a legitimate application to be communicated with**. This happens during deployment and does *not* necessarily mean the key is a JWT. 

tldr:
`JWT` = for communication between client and web servers
`AUTH_API_KEY` = doesn't really need to be JWT, I don't see how it is useful. It's just a way for the `auth` application to know the server is legitimate.

**Comment by nelsonic** [comment]
@LuchoTurtle thanks for your insightful reply. 👌 
Yeah, I was trying to rack my brain to figure out how using a `JWT` as the `AUTH_API_KEY` would be _useful_.
The only thing I can think of is validity checking on the side of the "consuming" app ...
A couple of days ago we saw the `client_id XYZ is not valid` error: https://github.com/dwyl/phoenix-chat-example/issues/150
![image](https://user-images.githubusercontent.com/194400/217873473-10998227-6060-4f37-9c75-14973d9eefe1.png)

To the person _using_ `auth_plug` this is not a very helpful error; a potential time-waster... ⏳ 💸 🔥 
If the `AUTH_API_KEY` was a `JWT`, we could easily add a `AuthPlug.verify_jwt/1` function
and thus avoid this class of errors completely. 
Obviously we can (_significantly_) improve our error handling/reporting in `auth` ...
but if there was a way of checking the `AUTH_API_KEY` with a function, 
it would _eliminate_ the need for handling this type of error 
because the engineer would immediately know if their `AUTH_API_KEY` is valid
and could even test it independently on https://jwt.io/ ✅ 

Hmm ... I'm still not "sold" on this in terms of cost-benefit. 🤔

**Comment by LuchoTurtle** [comment]
Thanks for the comprehensive explanation @SimonLab , makes much more sense now. I had some run-ins with this error but now knowing *how* it works helps a lot!
I'd open an issue but since `auth` is in the process of being rebuilt, I wager it's redundant. As @nelsonic said, "it's like re-arranging deck chairs on the Titanic 🚢"

**Comment by nelsonic** [comment]
@SimonLab thanks for stepping through the _existing_ code to give context.
I actually didn't _want_ to get bogged down in the _existing_ code 
because I've _already agreed_ it's over-complicated.
All I wanted to discuss was if `AUTH_API_KEY` should be a `JWT` or not. 
_However_ ... it has prompted me to open a follow-up question: https://github.com/dwyl/auth/issues/277

**Comment by LuchoTurtle** [comment]
Since some of this discussion was migrated to https://github.com/dwyl/auth/issues/277, should this issue be closed? The usage of the very existence of an `AUTH_API_KEY` was debated in https://github.com/dwyl/auth/issues/277.

Although it's confirmed in https://github.com/dwyl/auth/issues/277#issuecomment-1445010981 that we'll have an `AUTH_URL`/`AUTH_API_KEY`, it stated that it will have a `auth.domain.com/:person_id/:public_key` format, not JWT. 💭

**Comment by nelsonic** [comment]
Indeed. Not using `JWT` in `AUTH_API_KEY`. ✅ 
Thanks.
